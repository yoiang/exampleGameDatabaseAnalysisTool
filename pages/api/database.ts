import { NextApiRequest, NextApiResponse } from 'next'

import { google } from "googleapis";
import { googleSheetsRawAdapter } from "../../data/DatabaseRow";
import { DatabaseRow } from "../../models/DatabaseRow";

export default async (_req: NextApiRequest, _res: NextApiResponse): Promise<DatabaseRow[]> => {
  return getDatabase()
}

export const getDatabase = async (): Promise<DatabaseRow[]> => {
  try {
    const scopes = ["https://www.googleapis.com/auth/spreadsheets.readonly"];
    const jwt = new google.auth.JWT(
      process.env.GOOGLE_SHEETS_CLIENT_EMAIL,
      null,
      // we need to replace the escaped newline characters
      // https://stackoverflow.com/questions/50299329/node-js-firebase-service-account-private-key-wont-parse
      process.env.GOOGLE_SHEETS_PRIVATE_KEY.replace(/ \\n /g, "\n"),
      scopes
    );

    const sheets = google.sheets({ version: "v4", auth: jwt });

    const response = await sheets.spreadsheets.values.get({
      spreadsheetId: process.env.SPREADSHEET_ID,
      range: "Games",
    });

    const rows = response.data.values;
    return googleSheetsRawAdapter(rows)
  } catch (err) {
    // eslint-disable-next-line no-console
    console.log(err);
    // TODO: report to Sentry
  }

  return [];
}
