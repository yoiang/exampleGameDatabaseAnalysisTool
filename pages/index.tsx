import { GetStaticProps } from 'next'
import { Grommet } from 'grommet'
import { Tabs, Tab as GrommetTab } from 'grommet'

import { DatabaseRow } from '../models/DatabaseRow'
import { getDatabase } from './api/database'

import GlobalCSS from '../components/GlobalCSS'
import Head from '../components/Head'
import PivotTable from '../components/PivotTableSSR'
import RawTable from '../components/RawTable'
import { theme } from '../config/theme'
import Examples from '../components/Examples'

type StaticProps = {
  databaseRows: DatabaseRow[]
}

export type Tab = {
  label: string
  contents: JSX.Element
}

export const Home = ({ databaseRows }: StaticProps): JSX.Element => {
  const tabs: Tab[] = [
    {
      label: 'Pivot',
      contents: <PivotTable data={databaseRows} />,
    },
    {
      label: 'Examples',
      contents: <Examples data={databaseRows} />,
    },
    {
      label: 'Raw',
      contents: <RawTable rows={databaseRows} />,
    },
  ]

  return (
    <Grommet full theme={theme}>
      <Head pageTitle="Pivot Table" />

      <Tabs a11yTitle="Main navigation" alignControls="start" flex>
        {tabs.map((tab, index) => (
          <GrommetTab title={tab.label} key={`tab-${index}`}>
            {tab.contents}
          </GrommetTab>
        ))}
      </Tabs>

      <GlobalCSS />
    </Grommet>
  )
}

export const getStaticProps: GetStaticProps<StaticProps> = async (_context) => {
  const databaseRows = await getDatabase()

  return {
    props: {
      databaseRows,
    },
    // Next.js will attempt to re-generate the page:
    // - When a request comes in
    // - At most once every second
    revalidate: 1, // In seconds
  }
}

export default Home
