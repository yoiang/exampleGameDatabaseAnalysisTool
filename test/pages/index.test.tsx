import React from 'react'
import { render, fireEvent } from '../testUtils'
import { Home } from '../../pages/index'
import database from '../__mocks__/database'

describe('Home page', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(<Home databaseRows={database.getDatabase()} />, {})
    expect(asFragment()).toMatchSnapshot()
  })

  it('clicking button triggers alert', () => {
    const { getByText } = render(<Home databaseRows={database.getDatabase()}  />, {})
    window.alert = jest.fn()
    fireEvent.click(getByText('Test Button'))
    expect(window.alert).toHaveBeenCalledWith('With typescript and Jest')
  })
})
