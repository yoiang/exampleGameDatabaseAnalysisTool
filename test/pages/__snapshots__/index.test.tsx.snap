// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="jsx-2546281261 container"
  >
    <div
      position="static"
      style="color: rgba(0, 0, 0, 0.87); background-color: rgb(0, 188, 212); transition: all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms; box-sizing: border-box; font-family: Roboto, sans-serif; box-shadow: 0 1px 6px rgba(0, 0, 0, 0.12),
         0 1px 4px rgba(0, 0, 0, 0.12); border-radius: 0px; position: relative; z-index: 1100; width: 100%; display: flex; padding-left: 24px; padding-right: 24px;"
    >
      <button
        style="border: 10px; box-sizing: border-box; display: inline-block; font-family: Roboto, sans-serif; cursor: pointer; text-decoration: none; margin: 8px 8px 0px -16px; padding: 12px; outline: none; font-size: 0px; font-weight: inherit; position: relative; overflow: visible; transition: all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms; width: 48px; height: 48px; background: none;"
        tabindex="0"
        type="button"
      >
        <div>
          <svg
            style="display: inline-block; color: rgb(255, 255, 255); fill: #ffffff; height: 24px; width: 24px; user-select: none; transition: all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms;"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
            />
          </svg>
        </div>
      </button>
      <h1
        style="white-space: nowrap; overflow: hidden; text-overflow: ellipsis; margin: 0px; padding-top: 0px; letter-spacing: 0; font-size: 24px; font-weight: 400; color: rgb(255, 255, 255); height: 64px; line-height: 64px; flex: 1;"
      />
      <div
        style="box-sizing: border-box; background-color: rgb(232, 232, 232); height: 56px; padding: 0px 24px; display: flex; justify-content: space-between;"
      >
        <a
          href="/"
        >
          Home
        </a>
        <a
          href="/rawData"
        >
          Raw Data
        </a>
      </div>
    </div>
    <main
      class="jsx-2546281261"
    >
      0
    </main>
  </div>
</DocumentFragment>
`;
